networkx/convert_matrix.py

```diff
@@ -1170,83 +1171,45 @@ def to_numpy_array(
-    #
-    #   2) When working with multi(di)graphs, we must combine the values of all
-    #      edges between any two nodes in some manner.  This often takes the
-    #      form of a sum, min, or max.  Using the value 0 for a nonedge would
-    #      have undesirable effects with min and max, but using nanmin and
-    #      nanmax with initially nan values is not problematic at all.
-    #
-    # That said, there are still some drawbacks to this approach. Namely, if
-    # a real edge is nan, then that value is a) not distinguishable from
-    # nonedges and b) is ignored by the default combinator (nansum, nanmin,
-    # nanmax) functions used for multi(di)graphs. If this becomes an issue,
-    # an alternative approach is to use masked arrays.  Initially, every
-    # element is masked and set to some `initial` value. As we populate the
-    # graph, elements are unmasked (automatically) when we combine the initial
-    # value with the values given by real edges.  At the end, we convert all
-    # masked values to `nonedge`. Using masked arrays fully addresses reason 1,
-    # but for reason 2, we would still have the issue with min and max if the
-    # initial values were 0.0.  Note: an initial value of +inf is appropriate
-    # for min, while an initial value of -inf is appropriate for max. When
-    # working with sum, an initial value of zero is appropriate. Ideally then,
-    # we'd want to allow users to specify both a value for nonedges and also
-    # an initial value.  For multi(di)graphs, the choice of the initial value
-    # will, in general, depend on the combinator function---sensible defaults
-    # can be provided.
+    A = np.full((nlen, nlen), fill_value=nonedge, dtype=dtype, order=order)
+
+    # Corner cases: empty nodelist or graph without any edges
+    if nlen == 0 or G.number_of_edges() == 0:
+        return A
 
+    # Map nodes to row/col in matrix
+    idx = dict(zip(nodelist, range(nlen)))
+    if len(nodelist) < len(G):
+        G = G.subgraph(nodelist).copy()
+
+    # Collect all edge weights and reduce with `multigraph_weights`
     if G.is_multigraph():
-        # Handle MultiGraphs and MultiDiGraphs
-        A = np.full((nlen, nlen), np.nan, order=order)
-        # use numpy nan-aware operations
-        operator = {sum: np.nansum, min: np.nanmin, max: np.nanmax}
-        try:
-            op = operator[multigraph_weight]
-        except Exception as err:
-            raise ValueError("multigraph_weight must be sum, min, or max") from err
-
-        for u, v, attrs in G.edges(data=True):
-            if (u in nodeset) and (v in nodeset):
-                i, j = index[u], index[v]
-                e_weight = attrs.get(weight, 1)
-                A[i, j] = op([e_weight, A[i, j]])
-                if undirected:
-                    A[j, i] = A[i, j]
+        d = defaultdict(list)
+        for u, v, wt in G.edges(data=weight, default=1.0):
+            d[(idx[u], idx[v])].append(wt)
```

Comment:
While thinking about whether this could be vectorized, I learned that

```
np.full((10, 10), [], dtype=object)
```

does not work.  Would you agree that that is a bug?

Reply:
> Would you agree that that is a bug?

Yes, this does seem like a bug in NumPy, albeit for a very strange corner-case. I would have expected:

```python
>>> np.full((10, 10), [], dtype=object)
array([[list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])],
       [list([]), list([]), list([]), list([]), list([]), list([]),
        list([]), list([]), list([]), list([])]], dtype=object)
```

Reply:
@seberg Would this make for a good feature enhancement?  I guess it would require a change where numpy.full takes a factory as an argument?

Reply:
Well, there are two distinct things here, I guess?
* For `dtype=object` we could just store the list, that is a bit confusing, but a plausible (and possible) change.
* For lists specifically, you really need a factory, and it seems like you would need to add a `factor=` argument.  And on first sight, that might be a bit much? 

The only workaround/solution that I can think of right now is the following:
```
def factory():
    return []

factory = np.frompyfunc(factory, nin=0, nout=1)

result = factory(np.empty((10, 10), dtype=object))
```